src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -421,82 +424,190 @@ internal static void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<Re
+        }
+
+        // Abstracts away the type signature incompatibility between Memory and ReadOnlyMemory.
+        private interface IMemoryHandler<T>
+        {
+            int GetLength(in T memory);
+            MemoryHandle Pin(in T memory);
         }
 
+        private struct MemoryHandler : IMemoryHandler<Memory<byte>>
+        {
+            public int GetLength(in Memory<byte> memory) => memory.Length;
+            public MemoryHandle Pin(in Memory<byte> memory) => memory.Pin();
+        }
+
+        private struct ReadOnlyMemoryHandler : IMemoryHandler<ReadOnlyMemory<byte>>
+        {
+            public int GetLength(in ReadOnlyMemory<byte> memory) => memory.Length;
+            public MemoryHandle Pin(in ReadOnlyMemory<byte> memory) => memory.Pin();
+        }
+
+        // GCHandle.AddrOfPinnedObject performs some checks that are not
+        // necessary since we know we have pinned a specific SZArray.
+        private static unsafe IntPtr AddrOfPinnedArray<T>(T[] array) =>
+            (IntPtr) Unsafe.AsPointer(ref MemoryMarshal.GetArrayDataReference(array));
+
         // From https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfilescatter:
-        // "The file handle must be created with the GENERIC_READ right, and the FILE_FLAG_OVERLAPPED and FILE_FLAG_NO_BUFFERING flags."
+        // "The file handle must be created with [...] the FILE_FLAG_OVERLAPPED and FILE_FLAG_NO_BUFFERING flags."
         private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
             => handle.IsAsync && ((handle.GetFileOptions() & SafeFileHandle.NoBuffering) != 0);
 
-        private static async ValueTask<long> ReadScatterAtOffsetSingleSyscallAsync(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset, int totalBytes, CancellationToken cancellationToken)
+        // From the same source:
+        // "Each buffer must be at least the size of a system memory page and must be aligned on a system
+        // memory page size boundary. The system reads/writes one system memory page of data into/from each buffer."
+        // This method returns true if the buffers can be used by
+        // the Windows scatter/gather API, which happens when they are:
+        // 1. aligned at page size boundaries
+        // 2. exactly one page long each (our own requirement to prevent partial reads)
+        // 3. not bigger than 2^32 - 1 in total
+        // This function is also responsible for pinning the buffers if they
+        // are suitable and they must be unpinned after the I/O operation completes.
+        // The total size of the buffers is also returned.
+        private static bool TryPrepareScatterGatherBuffers<T, THandler>(IReadOnlyList<T> buffers,
+            THandler handler, [NotNullWhen(true)] out MemoryHandle[]? pinnedBuffers, out int totalBytes)
+            where THandler: struct, IMemoryHandler<T>
         {
+            int pageSize = Environment.SystemPageSize;
```

Comment:
Note that `Environment.SystemPageSize` is relatively slow today - it always calls the OS to get the page size.

Reply:
Thanks, I noticed it as well, I opened #57442 to track it but in the meantime I will cache the page size in a static readonly field.